### ISSUE TITLE
corpUser entities get randomly removed (status removed = true) from a datahub when stateful ingestion is True

### ISSUE THREAD

**Issue by bda618** [datahub-project/datahub#7934]
**Describe the bug**
I am the one who implemented a stateful ingestion for the LDAP ingestion. Recently we have found that sometimes LDAP accounts get randomly removed (status removed = true) when we execute LDAP ingestion on regular base. Essentially, a status aspect of some corpUser become set as "removed=true" without any reasonable reason. The current implementation for LDAP ingestion is mostly based on python helper classes and logic.

**To Reproduce**
The main issue that it is a random behavior and it is not clear how to reproduce this. We suspect this could happened under some external conditions, for example, when connection to LDAP server is broken and LDAP data for a particular request (page) are not reaching a checkpoint internal list storage, thus, they will be missing and later considered as "removed". In theory, this should be marked as a failure and hence a remove procedure should not be triggered. 

**Expected behavior**
corpUser entities shouldn't be removed from a datahub

**Screenshots**

**Desktop (please complete the following information):**
 - Datahub version [0.9.6.1]

**Additional context**
We have 400K ldap accounts and we run 7 ldap ingestion forks in parallel. 
I am looking for any help or advice from open source community to resolve (or even to reproduce) this issue.

**Comment by bda618** [comment]
I was able to reproduce this error and going to provide details here. 

Let's say you have 3000 employees and your LDAP configuration uses pagination with a page_size: 500. It means, there will be 6 pages with 500 employees for each response. You should also setup stateful_ingestion, enabled: True, because this is when "Clean up stale entities" will be activated. 

The first step to reproduce this error is to have a good ingestion run to generate and save checkpoint that will be used as a base (previous checkpoint) for the next run. On the second run, we are going to run ingestion a half way and then break the connection intentionally to emulate an external condition. In my case, I simply disconnected from a VPN and then restore connection by connecting back by using my VPN. 

Let's say we ingestion 1200 accounts and disconnected from a VPN. At this point ldap ingestion will put on hold. If I connect back, the rest of the accounts at the page (300) will be added to the state checkpoint (they kind of cached or something). Thus, I am going to have 1200 + 300 accounts at my current checkpoint. The previous run was a success, so it has 3000 corpUser accounts. Now, the current code has no any condition at gen_removed_entity_workunits() to check if any failures happened, so it will proceed and find a differences between current checkpoint (1500) and previous check point (3000). As result 1500 accounts will be soft deleted. 

What is also interesting is that NEW state will not be saved (because there is a condition for this). 
Also if the next run is a success, all soft deleted accounts will be restored. 
However, til the next run, some users of the datahub will not be able to use it because they were removed. 

I have tried a simple fix, by adding the below code to the top of gen_removed_entity_workunits() :
`        if len(self.source.get_report().failures) != 0:
            return`

and it fixed a problem. However, there could be a best solution.

**Comment by bda618** [comment]
@hsheth2, Thank you. I have applied my fix at my company temporarily, but in will rely on fix from your team in long term because it is part of a shared code for many different ingestion and I feel safe if you team will make an appropriate fix. 

Also, have a look at feat(ingest): https://github.com/datahub-project/datahub/pull/7784
It is related to LDAP only (specifically to LDAP manager).

**Comment by bda618** [comment]
This is still an issue and has to be addressed, this has been tested with version 9.6.1

**Comment by hsheth2** [comment]
@bda618 I made a fix that only emits soft deletes when the ingestion run succeeds in this PR https://github.com/datahub-project/datahub/pull/8111, so I believe this one should be largely fixed.

**Comment by bda618** [comment]
Git it, thank you for confirming.
